### DISCUSSION TITLE
Green tray icon even if disconnected

### DISCUSSION THREAD

**Discussion by kerotonic** [Martchus/syncthingtray#308] [General]
Hi,
I'm setting up syncthing to synchronize my devices (a desktop and two laptops, all working under Archlinux with an XFCE desktop). Syncthing is launched by systemd, with a system unit ([syncthing@user.service](mailto:syncthing@user.service)).

In order to have an easier view on syncthing I also installed syncthingtray via the Archlinux's AUR.

So far, everything seems to work correctly except for a (little) problem: when one of the devices gets disconnected, I expect the syncthingtray's icon (in the tray icons zone) to reflect that fact, by being greyed. Instead of that, it remains green.

What do I do or understand wrong ?

**Comment by Martchus** [comment]
The lastest release of Syncthing Tray shows the disconnected status only if Syncthing Tray itself is disconnected from Syncthing.

I recently implemented taking the status of remote devices into account, see https://github.com/Martchus/syncthingtray/issues/306. This feature hasn't been released yet. If you want to help me testing you can install `syncthingtray-git` from the AUR. You'll have to configure this in the connection settings to take effect. Note that this introduces a new status (icon with "X" badge). That status is reached in case no other remote devices are connected; so it is *not* reached if only one remote devices is disconnected but others are still connected.

**Comment by kerotonic** [comment]
Thank you for your answer.

I decided to give it a try. I'm trying to build syncthingtray-git but I have a problem when building one of its dependency: qtforkawesome-git. Here is the (relevant) log:

```keikoz@kdesktop ~/system-local/abs_build/AUR/qtforkawesome-git $ LANG=C makepkg -s
==> Making package: qtforkawesome-git 88.cef99f7-1 (Sat Dec 28 22:01:59 2024)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Updating qtforkawesome git repo...
  -> Found Fork-Awesome-1.2.0
==> Validating source files with sha256sums...
    qtforkawesome ... Skipped
    Fork-Awesome-1.2.0 ... Passed
==> Extracting sources...
  -> Creating working copy of qtforkawesome git repo...
Switched to a new branch 'makepkg'
rm: cannot remove '/home/keikoz/system-local/abs_build/AUR/qtforkawesome-git/src/Fork-Awesome-1.2.0': Is a directory
ln: failed to create symbolic link '/home/keikoz/system-local/abs_build/AUR/qtforkawesome-git/src/Fork-Awesome-1.2.0': File exists
==> Starting pkgver()...
==> Removing existing $pkgdir/ directory...
==> Starting build()...
CMake Error at /usr/share/qtutilities-git/cmake/qtutilities-gitConfig.cmake:80 (include):
  include could not find requested file:

    3rdParty
Call Stack (most recent call first):
  CMakeLists.txt:35 (find_package)
```

**Reply by Martchus** [reply]
Looks like c++utilities-git was not installed correctly in the environment where you attempted the build or incompatible versions of c++utilities-git and qtutilities-git  are installed. Simply make sure you have the latest version of both of these libraries installed and that you built c++utilities-git before qtutilities-git.

Not sure about these errors:

```
rm: cannot remove '/home/keikoz/system-local/abs_build/AUR/qtforkawesome-git/src/Fork-Awesome-1.2.0': Is a directory
ln: failed to create symbolic link '/home/keikoz/system-local/abs_build/AUR/qtforkawesome-git/src/Fork-Awesome-1.2.0': File exists
```

I haven't run into such a problem so far. Maybe you have to cleanup these files before attempting another build using whatever build script / AUR helper you're using.

**Reply by kerotonic** [reply]
The errors (that you paste) on qtforkawesome-git just come from the fact that I was running makepks -s for the second time in the same folder. Here is a "fresh" log of the problem:

```
keikoz@kdesktop ~/system-local/abs_build/AUR/qtforkawesome-git $ LANG=C makepkg -s
==> Making package: qtforkawesome-git 46.1c92ea5-3 (Sat Dec 28 22:21:29 2024)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Cloning qtforkawesome git repo...
Cloning into bare repository '/home/keikoz/system-local/abs_build/AUR/qtforkawesome-git/qtforkawesome'...
remote: Enumerating objects: 399, done.
remote: Counting objects: 100% (154/154), done.
remote: Compressing objects: 100% (93/93), done.
remote: Total 399 (delta 95), reused 107 (delta 57), pack-reused 245 (from 1)
Receiving objects: 100% (399/399), 70.93 KiB | 2.36 MiB/s, done.
Resolving deltas: 100% (250/250), done.
  -> Downloading Fork-Awesome-1.2.0...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 1336k    0 1336k    0     0  2223k      0 --:--:-- --:--:-- --:--:-- 9883k
==> Validating source files with sha256sums...
    qtforkawesome ... Skipped
    Fork-Awesome-1.2.0 ... Passed
==> Extracting sources...
  -> Creating working copy of qtforkawesome git repo...
Cloning into 'qtforkawesome'...
done.
  -> Extracting Fork-Awesome-1.2.0 with bsdtar
==> Starting pkgver()...
==> Updated version: qtforkawesome-git 88.cef99f7-1
==> Starting build()...
-- The C compiler identification is GNU 14.2.1
-- The CXX compiler identification is GNU 14.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at /usr/share/qtutilities-git/cmake/qtutilities-gitConfig.cmake:80 (include):
  include could not find requested file:

    3rdParty
Call Stack (most recent call first):
  CMakeLists.txt:35 (find_package)

[... leaving out the following...]

==> ERROR: A failure occurred in build().
    Aborting...
```

For c++utilities-git and qtutilities-git, I just used the AUR last versions (I just run git clone for each of them). 

```
local/c++utilities-git 1275.4d86993-1
local/qtutilities-git 569.fe5b3d9-1
```
c++utilities-git was indeed installed first (because anyway, qtutilities-git refused to build without it).

But still, qtforkawesome-git refuses to build. :)

**Reply by Martchus** [reply]
Ok, I could reproduce the issue. I simply haven't tried to rebuild qtforkawesome-git yet against the latest changes in c++utilities. I'm working on a fix.

**Reply by kerotonic** [reply]
Will try when you fix it. Thank you for your work, btw.

**Reply by Martchus** [reply]
I pushed a fix to c++utilities and it works for me now. You'll have to build both, c++utilities-git and qtutilities-git.

**Comment by kerotonic** [comment]
Ok finally I tried the -git version and it works. Now, this version doesn't really adresses (unless I miss something, which is highly possible) my initial problem (I had a little hope it would).

Just to be clear (which I'm not sure I was completely in my original post): my expectations wouldn't be that: if the device "A" is disconnected, this is reflected by greying out the tray icon on other devices (i.e. device "B"), but that: if the device "A" is disconnected from the internet/LAN (let's say, if I disable the ethernet iface), this is reflected by greying out the tray icon on this same devices (=device "A").

I had a hope that your development version (-git) could maybee address that: because if device A is disconnected from internet/LAN, it is also disconncted from all other devices, hence the icon would be greyed. But when I disconnect device A from my LAN, other (connected) devices still look as "idle" (is this intended ?).

To finish with this: in my sight, syncthingtray's main purpose is to give an information about syncthing's state and here, an important information is not given to the user (the fact that local syncthing instance is actually unable to sync because of disconnection). It's the kind of important information that is given by all equivalent softwares (pcloud, dropbox, synology drive client, and so on...).

Maybee could it be possible to add an additional "grey icon version", in addition to "paused" and "disconnected [from syncthing]", to reflect disconnection from the net ?

Hope I was clear and didn't miss anything...

**Reply by Martchus** [reply]
> I had a hope that your development version (-git) could maybee address that: because if device A is disconnected from internet/LAN, it is also disconncted from all other devices, hence the icon would be greyed. But when I disconnect device A from my LAN, other (connected) devices still look as "idle" (is this intended ?).

I would also hope it works this way :-)
Did you actually use the new configuration option? If so then the new feature is maybe not fully working yet. (I haven't spent much time testing it yet myself.)

**Reply by kerotonic** [reply]
> Did you actually use the new configuration option?

What option are you speaking about ?

**Reply by Martchus** [reply]
The "No remote device connected" checkbox in the connection settings introduced by commit https://github.com/Martchus/syncthingtray/commit/6b4eecf95c080e52aa36ea7188e73e26377b2c5c.

**Reply by kerotonic** [reply]
Yes it is checked. I restarted with a fresh configuration and it looks that by default, this option is already set.

**Reply by kerotonic** [reply]
Additional comment: I made some extra exploration. If i cut (from the net) all devices (except device "A"), device "A"'s icon becomes blue with a cross:
![Capture d’écran_2024-12-29_21-43-39](https://github.com/user-attachments/assets/c028fe76-2511-431b-a1f7-b3c23f3b88fd)
But if I leave other devices connected and just disconnect the device "A", syncthing doesn't updates the status of other devices (so device "B" is still shown as "up-to-date" in syncthing's UI and as "idle" in syncingtray) while theorically it is not possible to give that information. So, it looks coherent that syncingtray can't detect a disconnection of device "A" from the other pairs.

Now I still wonder: why not just rely on the unconnected status of the host (by checking if there is a valid network connection) to show a disconnected status for syncthing ?

**Reply by Martchus** [reply]
So I guess it actually works. The cross is supposed to indicate that the Syncthing instance is not connected to other instances. Perhaps I should also change the color to the disconnected color. 

For the status of you network connection I recommend to use a tray application tailored for that. Most platforms already come with such an icon out of the box. Note that also Syncthing is supposed to consider other devices as disconnected when there's no network connection but probably there's a grace period / timeout until that happens.

**Reply by kerotonic** [reply]
Ok, thank you for your answers and help.
